lib/bandit/websocket/frame.ex

```diff
@@ -95,18 +95,30 @@ defmodule Bandit.WebSocket.Frame do
   defp mask_and_length(length) when length <= 65_535, do: <<0::1, 126::7, length::16>>
   defp mask_and_length(length), do: <<0::1, 127::7, length::64>>
 
+  @mask_size 512
+
   # Note that masking is an involution, so we don't need a separate unmask function
-  def mask(payload, mask, acc \\ <<>>)
+  def mask(payload, mask) do
+    payload
+    |> do_mask(String.duplicate(<<mask::32>>, div(@mask_size, 32)), [])
+    |> IO.iodata_to_binary()
+  end
 
-  def mask(payload, mask, acc) when is_integer(mask), do: mask(payload, <<mask::32>>, acc)
+  defp do_mask(
+         <<h::unquote(@mask_size), rest::binary>>,
+         <<int_mask::unquote(@mask_size)>> = mask,
+         acc
+       ) do
+    do_mask(rest, mask, [acc, <<Bitwise.bxor(h, int_mask)::unquote(@mask_size)>>])
+  end
 
-  def mask(<<h::32, rest::binary>>, <<mask::32>>, acc) do
-    mask(rest, mask, acc <> <<Bitwise.bxor(h, mask)::32>>)
+  defp do_mask(<<h::32, rest::binary>>, <<int_mask::32, _mask_rest::binary>> = mask, acc) do
+    do_mask(rest, mask, [acc, <<Bitwise.bxor(h, int_mask)::32>>])
   end
 
-  def mask(<<h::8, rest::binary>>, <<current::8, mask::24>>, acc) do
-    mask(rest, <<mask::24, current::8>>, acc <> <<Bitwise.bxor(h, current)::8>>)
+  defp do_mask(<<h::8, rest::binary>>, <<current::8, mask::24, _mask_rest::binary>>, acc) do
+    do_mask(rest, <<mask::24, current::8>>, [acc, <<Bitwise.bxor(h, current)::8>>])
```

Comment:
@moogle19 I didn't bring over all the macro'd in matches here - as implemented here it just adds in an extra 512 bit stride (so we'll chew off 512 bits at a time until we have less than that left, then 32 bits at a time until we have less than that left, then 8 bits at a time). The difference in performance in negligible (as far as I can benchmark, anyway), and it's a bit easier for readers to grok.